### ISSUE TITLE
Trait resolution in associated types depends on ordering of traits in bounds.

### ISSUE THREAD

**Issue by Lymia** [rust-lang/rust#54296]
[The following code](https://play.rust-lang.org/?gist=accaf19c611e4c88c0fb590161e66cb5&version=stable&mode=debug&edition=2015) fails to compile:

```rust
use std::ops::*;
pub trait Trait {
    type Num: Sized + Shl<usize, Output = Self::Num> + Shl<u32, Output = Self::Num>;
}
pub fn function<T: Trait>(t: T::Num, b: usize) -> T::Num {
  t << b
}
```

It gives this error:

```
error[E0308]: mismatched types
 --> src/lib.rs:6:8
  |
6 |   t << b
  |        ^ expected u32, found usize

error: aborting due to previous error
```

This error is cleared if the declaration order of the two `Shl` bounds is reversed, as in [the following code](https://play.rust-lang.org/?gist=bf2dc4b7a4870ecb618bf8c6360ac651&version=stable&mode=debug&edition=2015):

```rust
use std::ops::*;
pub trait Trait {
    type Num: Sized + Shl<u32, Output = Self::Num> + Shl<usize, Output = Self::Num>;
}
pub fn function<T: Trait>(t: T::Num, b: usize) -> T::Num {
  t << b
}
```

This bug does not seem to happen when the bound is directly specified in the function, instead of through an associated type. In addition, explicitly calling `Shl::<usize>>::shl` works around the issue.

---------

In addition, there is what seems to be an related issue where trait resolution instead depends on the order that the traits are physically declared in. [This code](https://play.rust-lang.org/?gist=d2ae969bbba1a036d9a6899bf6331d32&version=stable&mode=debug&edition=2015) compiles properly:

```rust
use std::ops::*;

pub trait CheckedShl: Sized + Shl<u32, Output = Self> { }
pub trait PrimInt: Sized + Shl<usize, Output = Self> { }
pub trait Trait {
    type Num: PrimInt + CheckedShl;
}
pub fn function<T: Trait>(t: T::Num, b: usize) -> T::Num {
  t << b
}
```

But [this code](https://play.rust-lang.org/?gist=1ac713e88e96715c06302004225f0615&version=stable&mode=debug&edition=2015) doesn't:

```rust
use std::ops::*;

pub trait PrimInt: Sized + Shl<usize, Output = Self> { }
pub trait CheckedShl: Sized + Shl<u32, Output = Self> { }
pub trait Trait {
    type Num: PrimInt + CheckedShl;
}
pub fn function<T: Trait>(t: T::Num, b: usize) -> T::Num {
  t << b
}
```

**Comment by cuviper** [comment]
This happens all the way back to rustc 1.0.0!
(I didn't bother testing earlier...)

**Comment by cuviper** [comment]
In the related https://github.com/rust-num/num-traits/issues/87#issuecomment-422493337, this problem started happening after `rustfmt` reordered the module declarations, so the module with `PrimInt` now comes before the module with `CheckedShl`.  That's what @Lymia's "in addition" case is also showing.

When I was trying this, I found that I had to `cargo clean` too, otherwise a successful build would be "remembered" and still succeed in the failing order -- presumably something with incremental compilation.
